### PR TITLE
IdentifyApi Refactored

### PR DESCRIPTION
**Pull request by shnrndk**
This PR refactored the ```identifyApi()``` method. Now the api identification logic is taken from the ```canProcess()``` method to the newly created ```identifyApi()``` method. Previous ```identifiyApi()``` method was renamed to ```startProcess()``` method.
Related Issue:  https://github.com/wso2/api-manager/issues/1564

### CODE REVIEW

File: modules/core/src/main/java/org/apache/synapse/api/API.java [hunk 1]
```diff
@@ -276,20 +280,14 @@ public boolean canProcess(MessageContext synCtx) {
...
-                    (!ApiUtils.matchApiPath(path, context))) {
-                auditDebug("API context: " + context + " does not match request URI: " + path);
-                return false;
-            }
-
-            if(!versionStrategy.isMatchingVersion(synCtx)){
+            if (!ApiUtils.identifyApi(this, synCtx)) {
                 return false;
+            }else{
+                ApiUtils.getFullRequestPath(synCtx);
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/api/API.java:286]: can we simply move this out of the else block?

File: modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java [hunk 2]
```diff
@@ -158,6 +161,52 @@ public static String getSubRequestPath(MessageContext synCtx) {
         return (String) synCtx.getProperty(RESTConstants.REST_SUB_REQUEST_PATH);
     }
 
+    public static boolean identifyApi(API api, MessageContext synCtx) {
+        if (synCtx.getProperty(RESTConstants.IDENTIFIED_API) == null) {
+            String path = ApiUtils.getFullRequestPath(synCtx);
+            if (null == synCtx.getProperty(RESTConstants.IS_PROMETHEUS_ENGAGED) &&
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java:167]: It seems there is an issue with the original code here, what if PROMETEUS_ENGAGED is present and matchApiPath is false, we still have to return false; Let's fix that.

File: modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java [hunk 3]
```diff
@@ -158,6 +161,52 @@ public static String getSubRequestPath(MessageContext synCtx) {
...
+    /**
+     * Checks whether the provided resource is capable of processing the message from the provided message context.
+     * The resource becomes capable to do this when the it contains either the name of the api caller,
+     * or {@value ApiConstants#DEFAULT_BINDING_ENDPOINT_NAME}, in its binds-to.
+     *
+     * @param resource  Resource object
+     * @param synCtx    MessageContext object
+     * @return          Whether the provided resource is bound to the provided message context
+     */
+    public static boolean isBound(Resource resource, MessageContext synCtx) {
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java:191]: make it private

File: modules/core/src/main/java/org/apache/synapse/api/AbstractApiHandler.java [hunk 4]
```diff
@@ -51,19 +51,30 @@ protected boolean dispatchToAPI(Collection<API> apiSet, MessageContext synCtx) {
...
+            if (startProcess((API) apiObject, synCtx, defaultStrategyApiSet)) {
                 return true;
             }
         } else {
             //To avoid apiSet being modified concurrently
             List<API> duplicateApiSet = new ArrayList<API>(apiSet);
-            for (API api : duplicateApiSet) {
-                if (identifyAPI(api, synCtx, defaultStrategyApiSet)) {
+            API identifiedApi = (API) synCtx.getProperty(RESTConstants.IDENTIFIED_API);
+            //If normal path this goes to else, if logging enabled then goes inside the if method.
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/api/AbstractApiHandler.java:61]: This logic is not clear to me, can you please explain?

Comment (shnrndk) [reply modules/core/src/main/java/org/apache/synapse/api/AbstractApiHandler.java:61]: In the api logging flow in the first iteration it needs to identify the api. For that it calls the `identifyApi()` method. If the api is identified then that api gets stored in the IDENTIFIED_API property. Then in the second iteration, we do not need to identify the api again. Therefore we get the identified api from the from the IDENTIFIED_API property. So we do not need to iterate all the apis in the duplicateApiSet to identify the api.

But if the api logging is disabled then api is not identified previously, therefore IDENTIFIED_API property is null, So the api needs to be identified by iterating all the apis in the defaultStrategyApiSet.

File: modules/core/src/main/java/org/apache/synapse/api/API.java (hunk 1 above)
Comment (shnrndk) [reply modules/core/src/main/java/org/apache/synapse/api/API.java]: Fixed

File: modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java (hunk 2 above)
Comment (shnrndk) [reply modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java:167]: Added the new condition block, let me know if the debug log message needs to changed or not.

Comment (isudana) [reply modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java:167]: I still don't understand why we are checking Prometheus is enabled. Based on your logic we should be able to just get rid of the check. Shall we check the history we are checking this condition?

**Comment by arunans23** [comment]
Closing since the relevant changes are added from https://github.com/wso2/wso2-synapse/pull/2150

**Pull request closed without merging**